build-logic/packaging/src/main/kotlin/gradlebuild/packaging/transforms/ShrinkPublicApiClassesTransform.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gradlebuild.packaging.transforms
+
+import org.gradle.api.artifacts.transform.CacheableTransform
+import org.gradle.api.artifacts.transform.InputArtifact
+import org.gradle.api.artifacts.transform.TransformAction
+import org.gradle.api.artifacts.transform.TransformOutputs
+import org.gradle.api.artifacts.transform.TransformParameters
+import org.gradle.api.file.FileSystemLocation
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.SetProperty
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Input
+import org.gradle.internal.tools.api.ApiClassExtractor
+import org.gradle.internal.tools.api.impl.JavaApiMemberWriter
+import java.io.File
+import java.io.InputStream
+import java.util.Optional
+import java.util.zip.ZipEntry
+import java.util.zip.ZipFile
+
+
+/**
+ * Extract the public API classes from a jar file.
```

Comment:
I don't think we should include "public" in the name/doc here. This class can work for any sorts of API types, its the filter that determines what the API is. We could choose to only filter the private API with this class if we wanted to as long as we specified that as package list parameter.

Reply:
I adjusted this JavaDoc, but I think it's still make sense to include `public` here, since this `public` in in context of class API and not in a context of public packages. So we extract public API of a class. 

But I also renamed
```
@get:Input
val publicApiPackages: SetProperty<String>
```

to 
```
@get:Input
val packages: SetProperty<String>
```